File: packages-gen/dart/event/lib/base-box.dart
import 'package:xtalk_event/base-listener.dart' as event_common;

import 'package:xtalk_lang/common-data.dart' as xtd;

make_box(initial) {
  var initialFn = initial;
  if(!((initialFn.runtimeType).toString().contains("Function") || (initialFn.runtimeType).toString().contains("=>") || (initialFn).toString().startsWith("Closure"))){
    var initialData = initialFn;
    initialFn = (() {
      return initialData;
    });
  }
  var data = initialFn();
  return <dynamic, dynamic>{
    "::":"event.box",
    "listeners":<dynamic, dynamic>{},
    "data":data,
    "initial":initialFn
  };
}

check_event(event, path) {
  var evpath = event["path"];
  if(path.length > evpath.length){
    return false;
  }
  var arr_41344 = path;
  for(var i = 0; i < arr_41344.length; ++i){
    var v = arr_41344[i];
    if(v != evpath[i]){
      return false;
    }
  };
  return true;
}

add_listener(box, listener_id, path, callback, meta) {
  path = event_common.arrayify_path(path);
  return event_common.add_listener(box,listener_id,"box",callback,xtd.obj_assign(<dynamic, dynamic>{"box/path":path},meta),(event) {
    return check_event(event,path);
  });
}

var remove_listener = event_common.remove_listener;

var list_listeners = event_common.list_listeners;

get_data(box, path) {
  var data = box["data"];
  path = event_common.arrayify_path(path);
  return xtd.get_in(data,path);
}

set_data_raw(box, path, value) {
  var data = box["data"];
  if((() {
    var dart_truthy__41343 = xtd.arr_emptyp(path);
    return (null != dart_truthy__41343) && (false != dart_truthy__41343);
  })()){
    box["data"] = value;
  }
  else{
    return xtd.set_in(data,path,value);
  }
}

set_data(box, path, value) {
  var data = box["data"];
  path = event_common.arrayify_path(path);
  set_data_raw(box,path,value);
  return event_common.trigger_listeners(
    box,
    <dynamic, dynamic>{"path":path,"value":value,"data":data}
  );
}

del_data_raw(box, path) {
  path = event_common.arrayify_path(path);
  var data = box["data"];
  var ppath = path.sublist(0 - 0,path.length - 1);
  var parent = xtd.get_in(data,ppath);
  if(null != parent){
    var val = parent[path[path.length + -1]];
    parent.remove(path[path.length + -1]);
    return null != val;
  }
  return false;
}

del_data(box, path) {
  path = event_common.arrayify_path(path);
  var data = box["data"];
  if((() {
    var dart_truthy__41342 = del_data_raw(box,path);
    return (null != dart_truthy__41342) && (false != dart_truthy__41342);
  })()){
    return event_common.trigger_listeners(box,<dynamic, dynamic>{"path":path,"value":null,"data":data});
  }
}

reset_data(box) {
  var initial = box["initial"];
  return set_data(box,<dynamic>[],initial());
}

merge_data(box, path, value) {
  path = event_common.arrayify_path(path);
  var prev = get_data(box,path);
  var merged = xtd.obj_assign(xtd.obj_clone(prev),value);
  return set_data(box,path,merged);
}

append_data(box, path, value) {
  path = event_common.arrayify_path(path);
  var arr = xtd.arr_clone(get_data(box,path));
  arr.add(value);
  return set_data(box,path,arr);
}

File: packages-gen/js/libs/xt-event/base-box.js
const event_common = require("@xtalk/event/base-listener.js")

const xtd = require("@xtalk/lang/common-data.js")

function make_box(initial){
  let initialFn = initial;
  if(!("function" == (typeof initialFn))){
    let initialData = initialFn;
    initialFn = (function (){
      return initialData;
    });
  }
  let data = initialFn();
  return {
    "::":"event.box",
    "listeners":{},
    "data":data,
    "initial":initialFn
  };
}

function check_event(event,path){
  let evpath = event["path"];
  if(path.length > evpath.length){
    return false;
  }
  for(let i = 0; i < path.length; ++i){
    let v = path[i];
    if(v != evpath[i]){
      return false;
    }
  };
  return true;
}

function add_listener(box,listener_id,path,callback,meta){
  path = event_common.arrayify_path(path);
  return event_common.add_listener(box,listener_id,"box",callback,Object.assign({"box/path":path},meta),function (event){
    return check_event(event,path);
  });
}

var remove_listener = event_common.remove_listener;

var list_listeners = event_common.list_listeners;

function get_data(box,path){
  let {data} = box;
  path = event_common.arrayify_path(path);
  return xtd.get_in(data,path);
}

function set_data_raw(box,path,value){
  let {data} = box;
  if(xtd.arr_emptyp(path)){
    box["data"] = value;
  }
  else{
    return xtd.set_in(data,path,value);
  }
}

function set_data(box,path,value){
  let {data} = box;
  path = event_common.arrayify_path(path);
  set_data_raw(box,path,value);
  return event_common.trigger_listeners(box,{"path":path,"value":value,"data":data});
}

function del_data_raw(box,path){
  path = event_common.arrayify_path(path);
  let {data} = box;
  let ppath = path.slice(0,path.length - 1);
  let parent = xtd.get_in(data,ppath);
  if(null != parent){
    let val = parent[path[path.length + -1]];
    delete(parent[path[path.length + -1]]);
    return null != val;
  }
  return false;
}

function del_data(box,path){
  path = event_common.arrayify_path(path);
  let {data} = box;
  if(del_data_raw(box,path)){
    return event_common.trigger_listeners(box,{"path":path,"value":null,"data":data});
  }
}

function reset_data(box){
  let {initial} = box;
  return set_data(box,[],initial());
}

function merge_data(box,path,value){
  path = event_common.arrayify_path(path);
  let prev = get_data(box,path);
  let merged = xtd.obj_assign(xtd.obj_clone(prev),value);
  return set_data(box,path,merged);
}

function append_data(box,path,value){
  path = event_common.arrayify_path(path);
  let arr = get_data(box,path).slice();
  arr.push(value);
  return set_data(box,path,arr);
}

module.exports = {
  ["make_box"]:make_box,
  ["check_event"]:check_event,
  ["add_listener"]:add_listener,
  ["remove_listener"]:remove_listener,
  ["list_listeners"]:list_listeners,
  ["get_data"]:get_data,
  ["set_data_raw"]:set_data_raw,
  ["set_data"]:set_data,
  ["del_data_raw"]:del_data_raw,
  ["del_data"]:del_data,
  ["reset_data"]:reset_data,
  ["merge_data"]:merge_data,
  ["append_data"]:append_data
}